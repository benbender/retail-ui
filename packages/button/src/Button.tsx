import { Spinner } from '@retail-ui/spinner'
import { cloneElement } from '@retail-ui/utils'
import clsx from 'clsx'
import * as React from 'react'

import { ButtonStyles } from './styles'

type Ref = HTMLButtonElement
type ReactButtonProps = Omit<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  'disabled'
>

type ButtonColor = keyof typeof ButtonStyles['variant']['default']
type ButtonVariant = keyof typeof ButtonStyles['variant']
type ButtonSize = keyof typeof ButtonStyles['size']
type ButtonShape = keyof typeof ButtonStyles['shape']

export type ButtonProps = {
  size?: ButtonSize
  variant?: ButtonVariant
  shape?: ButtonShape
  color?: ButtonColor
  spinnerClassName?: string
  block?: boolean
  iconLeft?: React.ReactNode
  iconRight?: React.ReactNode
  isLoading?: boolean
  isDisabled?: boolean
  isBlock?: boolean
}

export const Button = React.forwardRef<Ref, ReactButtonProps & ButtonProps>(
  (props, ref) => {
    const {
      children,
      className,
      spinnerClassName,
      isDisabled,
      isBlock,
      isLoading,
      variant = 'default',
      color = 'primary',
      size = 'base',
      iconLeft,
      iconRight,
      shape,
      onClick,
      ...rest
    } = props

    const variantCls = ButtonStyles.variant

    const colorCls = variantCls[variant]
    const sizeCls = ButtonStyles.size
    const shapeCls = ButtonStyles.shape

    const cls = clsx(
      className,
      ButtonStyles.base,
      isDisabled && ButtonStyles.disabled,
      isBlock && ButtonStyles.block,
      isLoading && ButtonStyles.loading,
      colorCls[color],
      sizeCls[size],
      shape && shapeCls[shape],
    )
    // Spinner
    const spinnerSizeCls = ButtonStyles.spinner.size
    const spinnerCls = clsx('absolute', spinnerSizeCls[size], spinnerClassName)
    const isLoadingCls = isLoading ? 'opacity-0' : 'opacity-100'

    /* iconCls */
    const iconCls = ButtonStyles.icon
    const iconSizeCls = iconCls['size']

    const iconLeftCls = clsx(
      children && iconCls.variant.left,
      isLoadingCls,
      iconSizeCls[size],
    )
    const iconRightCls = clsx(
      children && iconCls.variant.right,
      isLoadingCls,
      iconSizeCls[size],
    )

    const iconLeftNode = cloneElement(iconLeft, { className: iconLeftCls })
    const iconRightNode = cloneElement(iconRight, { className: iconRightCls })

    return (
      <button
        ref={ref}
        className={cls}
        disabled={isDisabled}
        onClick={onClick}
        {...rest}
      >
        {isLoading && <Spinner className={spinnerCls} />}
        {iconLeftNode}
        <span className={isLoading ? 'opacity-0' : 'opacity-100'}>
          {children}
        </span>
        {iconRightNode}
      </button>
    )
  },
)

Button.displayName = 'Button'
