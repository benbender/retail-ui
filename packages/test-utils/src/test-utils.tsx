import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
