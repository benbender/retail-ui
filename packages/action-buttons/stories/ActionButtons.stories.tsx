import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
