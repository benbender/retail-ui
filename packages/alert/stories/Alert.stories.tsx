import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
