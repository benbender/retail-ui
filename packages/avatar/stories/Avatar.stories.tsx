import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
