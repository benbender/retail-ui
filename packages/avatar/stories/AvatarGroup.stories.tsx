import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
