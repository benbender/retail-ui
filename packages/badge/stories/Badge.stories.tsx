import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
