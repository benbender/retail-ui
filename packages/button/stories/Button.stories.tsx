import * as React from 'react'

import { Button, ButtonProps } from '../src/Button'

const meta = {
  title: 'Retail-UI/Button',
  component: Button,
}

export default meta

const Template = (args: ButtonProps) => (
  <Button {...args} data-testid={meta.title} />
)

export const Basic = Template.bind({})
// @ts-ignore
Basic.args = {
  children: 'Button',
}
