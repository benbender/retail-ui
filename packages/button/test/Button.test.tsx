import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Button from '../stories/Button.stories'

describe('Button', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Button.Basic />)
    expect(asFragment()).toMatchSnapshot()
  })
})
