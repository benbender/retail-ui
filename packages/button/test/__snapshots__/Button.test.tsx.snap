// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center leading-5 align-bottom transition-colors duration-150 ease-in transform scale-100 border cursor-pointer focus:outline-none dark:focus:shadow-outline-gray active:scale-95 overflow-hidden no-underline select-none text-white bg-purple-600 border-transparent active:shadow-outline-purple focus:bg-purple-700 hover:bg-purple-700 py-2 px-4 text-base font-medium rounded-lg"
    data-testid="Retail-UI/Button"
  >
    <span
      class="opacity-100"
    />
  </button>
</DocumentFragment>
`;
